internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func v1GetFiles(_ *Command, req *http.Request, _ *userState) Response {
+	query := req.URL.Query()
+	action := query.Get("action")
+	switch action {
+	case "read":
+		paths := query["path"]
+		if len(paths) == 0 {
+			return statusBadRequest("must specify one or more paths")
+		}
+		return readFilesResponse{paths: paths}
+	case "list":
+		pattern := query.Get("pattern")
+		if pattern == "" {
+			return statusBadRequest("must specify pattern")
+		}
+		directory := query.Get("directory")
+		if directory != "true" && directory != "false" && directory != "" {
+			return statusBadRequest(`directory parameter must be "true" or "false"`)
+		}
+		return listFiles(pattern, directory == "true")
+	default:
+		return statusBadRequest("invalid action %q", action)
+	}
+}
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+	f     *os.File
+}
+
+// Custom Response implementation to serve the multipart.
+type readFilesResponse struct {
+	paths []string
+}
+
+// Reading files
+
+func (r readFilesResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.Header.Get("Accept") != "multipart/form-data" {
+		errResp := statusBadRequest(`must accept multipart/form-data`)
+		errResp.ServeHTTP(w, req)
+		return
+	}
+
+	// We open all the files first so we can detect not-found and
+	// permission-denied errors for each file, as we send the metadata and
+	// errors first.
+	result := make([]fileResult, len(r.paths))
+	var firstErr error
+	for i, p := range r.paths {
+		f, err := openForRead(p)
+		if err != nil {
+			if firstErr == nil {
+				firstErr = err
+			}
+		} else {
+			defer f.Close()
+		}
+		result[i] = fileResult{
+			Path:  p,
+			Error: fileErrorToResult(err),
+			f:     f,
+		}
+	}
+
+	// Write HTTP status and headers.
+	mw := multipart.NewWriter(w)
+	header := w.Header()
+	header.Set("Content-Type", mw.FormDataContentType())
+	status := http.StatusOK
+	if firstErr != nil {
+		status = http.StatusBadRequest
+	}
+	w.WriteHeader(status)
+
+	// Write first part: response metadata in JSON format.
+	mh := textproto.MIMEHeader{}
+	mh.Set("Content-Type", "application/json")
+	mh.Set("Content-Disposition", `form-data; name="response"`)
+	part, err := mw.CreatePart(mh)
+	if err != nil {
+		http.Error(w, "\n"+err.Error(), http.StatusInternalServerError)
```

Comment:
Yeah, this is tricky. By the time you know you have an error, you've already written the headers and some bytes in the body. So with a streamed multipart response like this there's no good way to include all the errors up front (unless you do something nasty like read the entire files' contents into memory).

Your idea with a list of errors at the end is good ... I started implementing that, but then there were two pieces of metadata, the first one with just the paths, and the last with any additional errors. Then I realized, why not just put the entire "response" JSON metadata at the end *instead* of the start? This actually works out really nicely, and allows you to include all errors. The HTTP status is always 200 OK, and the metadata (really just paths and errors) are always in the `response` field at the end. It's simpler code and more consistent. Let me know what you think.

However, there are still a couple of cases where you have to stop early -- if I get an error outputting the next multipart part, I still return early with a broken response. I don't see a real workaround for this (it's kind of like getting an error when you're logging an error).

The "\n" prefix in the early returns are just to put the error message on a new line, so that if a human is examining the output, the error message will be sure to be on a line of its own.

Reply:
The StatusInternalServerError there is not actually a 200, but it won't be sent as the actual HTTP status as is generally the intention of the http.Errror function, because the header has been sent already. Then, we prefix the error message with \n to dump an error message in the middle of some bulk data in the hope someone will introspect the output to see why it was actually wrong, instead of planning for an error to be reported properly to the other side. So we're abusing a function that clearly is not meant for this, and the result is not even so good.

I think we need to have a convention that makes building the server and clients straightforward, while following a nice convention that is easy to explain, and hopefully compatible with alternative clients that are not our own.

After going back and forth on this a few times, I think we should go with your suggestion of having "response" last and including both any reading errors and the digest computed on-the-fly (so we only read files once). On the write side we then store all files temporary in the local filesystem, evaluate the "response" to make sure we're happy with the entire operation, and only then commit the modifications to the filesystem.

What made me tilt towards this is the realization that today the mime parser in Go will do that itself when we attempt to parse a message that is over 10MB, which means any streaming optimizations we do are pointless, at least for the time being and without someone doing work to improve this. So we may as well just do that.

As part of reading about this, I also found the answer to the other question: the way browsers send the filename is indeed via an independent filename="..." field, with a form-data such as name="files" that is repeated for all the files present in the given form field (for our case, all of them). The header entry for each file looks like:
```
Content-Disposition: form-data; name="files"; filename="dir/name/somefile.txt"
```

From [section 4.2 of RFC7578](https://tools.ietf.org/html/rfc7578#section-4.2):

>    For form data that represents the content of a file, a name for the file SHOULD be supplied as well, by using a "filename" parameter of the Content-Disposition header field.  The file name isn't mandatory for cases where the file name isn't available or is meaningless or private; this might result, for example, when selection or drag-and- drop is used or when the form data content is streamed directly from a device.
> 
> If a "filename" parameter is supplied, the requirements of Section 2.3 of [RFC2183] for the "receiving MUA" (i.e., the receiving Mail User Agent) apply to receivers of multipart/form-data as well: do not use the file name blindly, check and possibly change to match local file system conventions if applicable, and do not use directory path information that may be present.

From [section 4.3 of RFC7578](https://tools.ietf.org/html/rfc7578#section-4.3):

> To match widely deployed implementations, multiple files MUST be sent by supplying each file in a separate part but all with the same "name" parameter.

From [section 2.3 of RFC2183](https://tools.ietf.org/html/rfc2183#section-2.3):
> The receiving MUA SHOULD NOT respect any directory path information that may seem to be present in the filename parameter.  The filename should be treated as a terminal component only.  Portable specification of directory paths might possibly be done in the future via a separate Content-Disposition parameter, but no provision is made for it in this draft.

We're clearly not following the RFC purely as intended, since we use an extra field with data, and we plan to respect the directory to define the target location, but I can't see a reason for us to not approach the standard design as much as possible and use the filename field somewhat as planned. The Go package supports it already as well, and given the widespread use I assume Python does too.

Reply:
You're right that my use of `http.Error` with `StatusInternalServerError` was misleading here. I've changed it to just `fmt.Fprint` to the output writer in those cases. Note that for these few cases (writing and encoding the metadata part) we can't do much better than printing to the output and hoping it succeeds. In fact, `CreatePart` basically only returns an error if there was an error writing to the output, so it may not do much good to write more anyway. What will happen in practice in the rare cases that this does occur is the client will not find the multipart boundary and will fail due to invalid multipart / unexpected EOF when reading the multipart data.

Thanks for the links to the multipart RFC. I had thought about using the filename field earlier, but originally decided against it because I know some clients strip off the path part of it (per the "SHOULD NOT" warning in RFC section 2.3 that you've quoted). However, Go does not, and we'll have a lot of control over the Python client (there's no multipart parser in the stdlib) so I think it's okay. I've updated the code to use "filename" -- it is simpler and cleaner.

Reply:
Oh, I forgot to mention: I did see your note about calculating the digest as we read and including it in the metadata at the end, and I think that's a good idea, however, we'd really love to get this first PR included in Juju for tomorrow's 2.9rc10 release (we're trying to do it on a weekly cadence). I believe that's something that can easily be added later, and clients can check it if it's present.